pymaker/join.py

```diff
@@ -0,0 +1,120 @@
+# This file is part of Maker Keeper Framework.
+#
+# Copyright (C) 2019-2021 EdNoepel
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+
+from web3 import Web3
+
+from pymaker import Address, Contract, Transact
+from pymaker.ilk import Ilk
+from pymaker.token import DSToken, ERC20Token
+from pymaker.numeric import Wad, Ray, Rad
+
+
+logger = logging.getLogger()
+
+
+class Join(Contract):
+    def __init__(self, web3: Web3, address: Address):
+        assert isinstance(web3, Web3)
+        assert isinstance(address, Address)
+
+        self.web3 = web3
+        self.address = address
+        self._contract = self._get_contract(web3, self.abi, address)
+        self._token: DSToken = None
+
+    def approve(self, approval_function, source: Address):
```

Comment:
Not pressing, but should `approve` and `approve_token` be moved directly into `Contract` since it appears to already be generalized, and is otherwise duplicated?

Reply:
Maybe.  It seems to rely on the contract having some pieces of ABI shared with ERC20 tokens.  As such we could create an abstract class named `ContractWithERC20LikeApprovals`.  Or have I misinterpreted that?

Reply:
Good point, yeah there are contracts which don't handle token transfers and wouldn't involve approvals so likely wouldn't be appropriate to put into the base `Contract` class. Something like `ContractWithERC20LikeApprovals` might be valuable in the future, but would say that's out of scope given the extent of the current changes.